### PR TITLE
TF-2520 Ability to print a mail on Web

### PR DESCRIPTION
**Pull request by dab246**
## Issue

#2520 

## Resolved

- Chrome:

https://github.com/linagora/tmail-flutter/assets/80730648/bc634670-a088-411a-ba91-08d1ddf4ca86

- Firefox:

https://github.com/linagora/tmail-flutter/assets/80730648/06f35e4b-8ab7-4934-817d-5443ebf9107c

- Safari:

https://github.com/linagora/tmail-flutter/assets/80730648/666ea656-2dd0-4711-93e4-5f9e7a0d4b93

### CODE REVIEW

**Review by hoangdat — APPROVED** [review]

**Comment by hoangdat** [comment]
please rebase @dab246

**Comment by hoangdat** [comment]
image

[[GitHub] Your Dependabot alerts for the week of Sep 13 - Sep 20 - GitHub <noreply@github.com> - 2022-09-20 0913.eml.txt](https://github.com/linagora/tmail-flutter/files/14446459/GitHub.Your.Dependabot.alerts.for.the.week.of.Sep.13.-.Sep.20.-.GitHub.noreply%40github.com.-.2022-09-20.0913.eml.txt)

**Comment by hoangdat** [comment]
```
[TwakeMail] CacheExceptionThrower::throwException():error: Instance of 'TooManyPagesException' | stackTrace: ../dart-sdk/lib/_internal/js_dev_runtime/private/ddc_runtime/errors.dart 294:3  throw_
../packages/pdf/src/widgets/multi_page.dart 251:9                               generate
../packages/pdf/src/widgets/document.dart 117:9                                 addPage
../packages/core/utils/print_utils.dart 342:18                                  <fn>
../dart-sdk/lib/_internal/js_dev_runtime/patch/async_patch.dart 45:50           <fn>
../dart-sdk/lib/async/zone.dart 1407:47                                         _rootRunUnary
../dart-sdk/lib/async/zone.dart 1308:19                                         runUnary
../dart-sdk/lib/async/future_impl.dart 162:18                                   handleValue
../dart-sdk/lib/async/future_impl.dart 846:44                                   handleValueCallback
../dart-sdk/lib/async/future_impl.dart 875:13                                   _propagateToListeners
../dart-sdk/lib/async/future_impl.dart 647:5                                    [_completeWithValue]
../dart-sdk/lib/async/future_impl.dart 721:7                                    <fn>
../dart-sdk/lib/async/zone.dart 1399:13                                         _rootRun
../dart-sdk/lib/async/zone.dart 1301:19                                         run
../dart-sdk/lib/async/zone.dart 1209:7                                          runGuarded
../dart-sdk/lib/async/zone.dart 1249:23                                         callback
../dart-sdk/lib/async/schedule_microtask.dart 40:11                             _microtaskLoop
../dart-sdk/lib/async/schedule_microtask.dart 49:5                              _startMicrotaskLoop
../dart-sdk/lib/_internal/js_dev_runtime/patch/async_patch.dart 181:7           <fn>

[TwakeMail] BaseController::_performFilterExceptionInError(): Instance of 'TooManyPagesException'
[TwakeMail] BaseController::onData:exception: null
[TwakeMail] BaseController::handleFailureViewState(): PrintEmailFailure
```

**Review by tddang-linagora — APPROVED** [review]

**Comment by hoangdat** [comment]
### Context:
- can not levarage `htmltopdfwidget`
  - pdf page size smaller than pdf widget
  - some tag not supported

### Solution:
- create a html with:
  - icon app - app name
  - date time
  - sender, recipients
  - subject
  - ...
  - content
  - attachment
- create object -> use `js` to print

**Comment by dab246** [comment]
@hoangdat  Please review again

**Comment by hoangdat** [comment]
@chibenwa please review
Some notes:
- how about to print the bluebar?
image
- how about the time in printed pdf (UTC or the same with app)?

**Comment by chibenwa** [comment]
> how about to print the bluebar?

Ok but I'm ok to also print without the bkue bar...

> how about the time in printed pdf (UTC or the same with app)?

Please same than the app

File: core/lib/presentation/utils/html_transformer/dom/remove_tag_hidden_transformers.dart [hunk 1]
```diff
@@ -0,0 +1,22 @@
+import 'package:core/data/network/dio_client.dart';
+import 'package:core/presentation/utils/html_transformer/base/dom_transformer.dart';
+import 'package:html/dom.dart';
+
+class RemoveTagHiddenTransformer extends DomTransformer {
```
Comment (hoangdat) [inline core/lib/presentation/utils/html_transformer/dom/remove_tag_hidden_transformers.dart:5]: remove it

Comment (dab246) [reply core/lib/presentation/utils/html_transformer/dom/remove_tag_hidden_transformers.dart:5]: done

File: core/lib/utils/print_utils.dart [hunk 2]
```diff
@@ -0,0 +1,247 @@
...
+import 'package:html/dom.dart';
+import 'package:html/parser.dart';
+
+class PrintUtils {
+  final ImagePaths _imagePaths;
+  final FileUtils _fileUtils;
+
+  PrintUtils(this._imagePaths, this._fileUtils);
+
+  Future<Element> _createUserInformationElement({
```
[the remark this answers is not in the document]
Comment (dab246) [reply core/lib/utils/print_utils.dart:19]: done

File: lib/features/email/domain/usecases/print_email_interactor.dart [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+import 'package:tmail_ui_user/features/email/domain/state/print_email_state.dart';
+
+class PrintEmailInteractor {
+  final EmailRepository emailRepository;
+
+  PrintEmailInteractor(this.emailRepository);
+
+  Stream<Either<Failure, Success>> execute(EmailPrint emailPrint) async* {
+    try {
+      yield Right(PrintEmailLoading());
```
Comment (hoangdat) [inline lib/features/email/domain/usecases/print_email_interactor.dart:17]: We should handle it in UI by showing toast.

File: lib/features/email/domain/usecases/print_email_interactor.dart [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+      final htmlContentTransformed = await _transformHtmlEmailContent(emailPrint.emailContent);
+      final newEmailPrint = emailPrint.fromEmailContent(htmlContentTransformed);
+      await emailRepository.printEmail(newEmailPrint);
+      yield Right(PrintEmailSuccess());
+    } catch (e) {
+      yield Left(PrintEmailFailure(exception: e));
```
Comment (hoangdat) [inline lib/features/email/domain/usecases/print_email_interactor.dart:23]: also showing toast for this case

File: lib/features/email/domain/usecases/print_email_interactor.dart (hunk 3 above)
Comment (dab246) [reply lib/features/email/domain/usecases/print_email_interactor.dart:17]: done

File: lib/features/email/domain/usecases/print_email_interactor.dart (hunk 4 above)
Comment (dab246) [reply lib/features/email/domain/usecases/print_email_interactor.dart:23]: done

File: lib/main/localizations/app_localizations.dart [hunk 5]
```diff
@@ -3819,4 +3819,18 @@ class AppLocalizations {
       name: 'replyToEmailAddressPrefix',
     );
   }
+
+  String get printYourEmailHasStarted {
+    return Intl.message(
+      'Print your email has started',
```
Comment (hoangdat) [inline lib/main/localizations/app_localizations.dart:3825]: `Printing in progress`

File: lib/main/localizations/app_localizations.dart [hunk 6]
```diff
@@ -3806,4 +3806,31 @@ class AppLocalizations {
...
+      name: 'printYourEmailHasStarted'
+    );
+  }
+
+  String get printYourEmailHasFailed {
+    return Intl.message(
+      'Print your email has failed',
```
Comment (hoangdat) [inline lib/main/localizations/app_localizations.dart:3832]: Printing failed

File: lib/main/localizations/app_localizations.dart (hunk 5 above)
Comment (dab246) [reply lib/main/localizations/app_localizations.dart]: done

File: lib/main/localizations/app_localizations.dart (hunk 6 above)
Comment (dab246) [reply lib/main/localizations/app_localizations.dart]: done

**Review by hoangdat — APPROVED** [review]

**Pull request merged**